pytorch_lightning/callbacks/swa.py

```diff
@@ -0,0 +1,234 @@
+
+        if len(lr_schedulers) > 1:
+            raise MisconfigurationException("SWA currently not supported for more than 1 lr_scheduler.")
+
+        self._max_epochs = trainer.max_epochs
+
+        # convert float to integer.
+        if isinstance(self._swa_epoch_start, float):
+            self._swa_epoch_start = int(self._max_epochs * self._swa_epoch_start)
+
+        self._model_contains_batch_norm = self.pl_module_contains_batch_norm(pl_module)
+
+        if self._model_contains_batch_norm:
+            # virtually increase max_epochs to perform batch norm update on latest epoch.
+            trainer.max_epochs += 1
+
+    def on_train_epoch_start(self, trainer, pl_module):
+        if trainer.current_epoch == self._swa_epoch_start:
+            optimizers = trainer.optimizers
+            lr_scheduler = trainer.lr_schedulers[0]["scheduler"]
+
+            self._swa_scheduler = SWALR(
+                optimizers[0],
+                swa_lr=self._swa_lrs,
+                anneal_epochs=self._annealing_epochs,
+                anneal_strategy=self._annealing_strategy,
+                last_epoch=trainer.max_epochs if self._annealing_strategy == "cos" else -1
+            )
+
+            rank_zero_warn(f"swapping lr_scheduler {lr_scheduler} for {self._swa_scheduler}")
+
+            trainer.lr_schedulers[0]["scheduler"] = self._swa_scheduler
+
+            self.n_averaged = torch.tensor(0, dtype=torch.long, device=pl_module.device)
+
+        elif self._model_contains_batch_norm and trainer.current_epoch == self._max_epochs:
+            # Turn off backwards and optimization, only update batch norm stats
+            self.transfer_weights(self._average_model, pl_module)
+
+            # perform accumulation over the entire train_dataloader
+            self._accumulate_grad_batches = trainer.accumulate_grad_batches
+            trainer.accumulate_grad_batches = len(trainer.train_dataloader)
+
+        if trainer.current_epoch >= self._swa_epoch_start:
+            self.update_parameters(self._average_model, pl_module, self.n_averaged, self.avg_fn)
+
+    @staticmethod
+    def update_parameters(average_model, model, n_averaged, avg_fn):
+        """
+        Credit to PyTorch Team.
+        Taken from https://github.com/pytorch/pytorch/blob/v1.7.1/torch/optim/swa_utils.py#L103
+        """
+        for p_swa, p_model in zip(average_model.parameters(), model.parameters()):
+            device = p_swa.device
+            p_model_ = p_model.detach().to(device)
+
+            if n_averaged == 0:
+                p_swa.detach().copy_(p_model_)
+            else:
+                p_swa.detach().copy_(
+                    avg_fn(
+                        p_swa.detach(),
+                        p_model_,
+                        n_averaged.to(device)
+                    )
+                )
+        n_averaged += 1
+
+    @staticmethod
+    def transfer_weights(average_module, pl_module):
```

Comment:
this shall be also applied in `on_fit_end` right?

Reply:
No need.

Reply:
ok, but how do you update after the last epoch?